cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/io/cdap/cdap/etl/batch/preview/LimitingRecordReader.java

```diff
@@ -29,47 +31,77 @@
  * @param <V> type of value to read
  */
 public class LimitingRecordReader<K, V> extends RecordReader<K, V> {
-  private final RecordReader<K, V> delegate;
-  private final int maxToRead;
+
+  private final InputFormat<K, V> delegateFormat;
+  private TaskAttemptContext context;
+  private LimitingInputSplit inputSplit;
+  private int perSplitLimit;
   private int numRead;
+  private Iterator<InputSplit> splitIterator;
+  private RecordReader<K, V> currentReader;
 
-  public LimitingRecordReader(RecordReader<K, V> delegate, int maxToRead) {
-    this.delegate = delegate;
-    this.maxToRead = maxToRead;
+  LimitingRecordReader(InputFormat<K, V> delegateFormat) {
+    this.delegateFormat = delegateFormat;
   }
 
   @Override
   public void initialize(InputSplit split, TaskAttemptContext context) throws IOException, InterruptedException {
-    delegate.initialize(split, context);
+    if (!(split instanceof LimitingInputSplit)) {
+      throw new IOException("Expected input split class " + LimitingInputSplit.class.getName()
+                              + ", but got " + split.getClass().getName());
+    }
+    this.context = context;
+    this.inputSplit = (LimitingInputSplit) split;
+    this.splitIterator = inputSplit.getInputSplits().iterator();
+
+    // Round up the per split limit so that each reader at most is opened once
+    int numberOfSplits = inputSplit.getInputSplits().size();
+    int perSplitLimit = (inputSplit.getRecordLimit() + numberOfSplits - 1) / numberOfSplits;
```

Comment:
Though the current behavior is a limit per split, don't we actually want the limit to be applied overall?

Reply:
We calculate limit per split from the overall limit. And in the nextKeyValue() method, the overall limit is enforced there.

Reply:
Also there is always only one LimitingInputSplit. That's why the read count can be tracked locally.